Provide option to separate self-evaluation from comparison period, and to add custom self-eval instructions

wynnset opened the pull request:
- On assignment edit screen, add checkbox and inputs to allow user to enter specific start/end dates for self-eval
- On assignment edit screen, add new text field to allow for entering custom instructions
- On assignment edit screen, update validation of dates based on new logic
- On course / assignment view screen, update metadata, buttons, and help text, to reflect new changes
- If self-eval period is explicitly defined, allow for self-eval anytime during the self-eval period, even if comparisons not done
- If self-eval period is explicitly defined, do not forward students to self-eval screen after comparisons done

Closes #687, #764, #793

andrew-gardener commented on compair/models/assignment.py:
```diff
@@ -168,6 +172,50 @@ def after_comparing(self):
 
+    @hybrid_property
+    def self_eval_period(self):
+        now = dateutil.parser.parse(datetime.datetime.utcnow().replace(tzinfo=pytz.utc).isoformat())
```
Think this is slightly off
```
now = dateutil.parser.parse(datetime.datetime.utcnow().replace(tzinfo=pytz.utc).isoformat())
if not self.enable_self_evaluation:
    return False
elif not self.self_eval_start:
    compare_start = self.compare_start.replace(tzinfo=pytz.utc)
    answer_end = self.answer_end.replace(tzinfo=pytz.utc)
    if self.compare_start:
        return compare_start <= now
    else:
        return answer_end <= now
else:
    return self.self_eval_start.replace(tzinfo=pytz.utc) <= now < self.self_eval_end.replace(tzinfo=pytz.utc)
```

wynnset replied:
Thanks, good idea to add in `enable_self_evaluation`. I think this better covers it, though: 
```python
        if not self.enable_self_evaluation:
            return False
        elif not self.self_eval_start:
            return self.compare_period
        else:
            return self.self_eval_start.replace(tzinfo=pytz.utc) <= now < self.self_eval_end.replace(tzinfo=pytz.utc)
```

andrew-gardener replied:
If you use compare period, then self evaluations will close when comparisons close. Not sure if this is desired. @lenglund thoughts?

lenglund replied:
In general, the request from the instructor side is for self-evals to exist independent of all other work being done (the thought being students submit and compare and read feedback first, then have time to reflect / self-eval later). Does that answer the question?

wynnset replied:
This is the current functionality. If the instructor wants to give students that extra flexibility, they can explicitly set the self-eval dates, in which case the self-eval is not affected by comparison end date. If the self-eval dates are not set, however, I think it makes sense to keep the current logic. (I thought that's what we had decided already?)

andrew-gardener replied:
I guess my only hold up is that if there is no self-eval date set, then it shouldn't ever end (whether there is a comparison period or not). That's how comparisons handle having no dates set

wynnset replied:
But that is not the logic currently. Currently, if there is no self-eval date set, self-eval ends at the same time as the comparison end. Instructors still will have the option to set the self-eval date to go until the end of the course by setting it explicitly. I am thinking maybe something we can do is default the self-eval end date to the course end date (currently it's not set), so all they would have to do is check the box and it will set it for them. 

At the end of the day, I think this just comes down to a decision. Do we want to change the existing logic even when they are not setting self-eval dates?

andrew-gardener commented on compair/models/assignment.py:
```diff
@@ -168,6 +172,50 @@ def after_comparing(self):
+            return self.self_eval_start.replace(tzinfo=pytz.utc) <= now < grace
+
+    @hybrid_property
+    def after_self_eval(self):
```
not being used right now so I think its safe to remove for now

andrew-gardener commented on compair/models/assignment.py:
```diff
@@ -68,6 +71,7 @@ class Assignment(DefaultTableMixin, UUIDMixin, ActiveMixin, WriteTrackingMixin):
 
     # hybrid and other functions
     course_uuid = association_proxy('course', 'uuid')
+    course_end_date = association_proxy('course', 'end_date')
```
can be removed since we only need this in the time period validator

wynnset replied:
More defensive programming - removing.

andrew-gardener commented on compair/static/less/assignment.less:
```diff
@@ -22,18 +22,14 @@
     padding-top: .5em;
     color: @breadcrumb-gray;
     border-top: 2px dotted @green-blue;
+    overflow-x: hidden;
```
What was overflowing? (haven't run the code yet so not sure what this is hiding)

wynnset replied:
It's just a CSS "trick" to add separators between items that may span multiple lines without having the separator show at the beginning (where there should be no separators). This particular directive hides the separator at the beginning (along with a negative margin on the actual item).

wynnset replied:
It just looks much cleaner this way compared to the way it was looking before.

andrew-gardener commented on compair/static/modules/assignment/assignment-form-partial.html:
```diff
@@ -334,6 +330,81 @@ <h1 ng-if="method == 'copy'">Duplicate: {{originalAssignment.name}}</h1>
+    <fieldset>
+        <legend>Self-Evaluation Settings</legend>
+        <div class="instructional-text">
+            <p>If enabled, students will also be required to self-evaluate by providing feedback on their own answer (after answering and comparing).</p>
```
`(after answering and comparing)` isn't necessarily true (when date set). not sure how to reword better though

wynnset replied:
I'll just remove "after comparing"

andrew-gardener commented on compair/static/modules/assignment/assignment-form-partial.html:
```diff
@@ -334,6 +330,81 @@ <h1 ng-if="method == 'copy'">Duplicate: {{originalAssignment.name}}</h1>
+
+            <div class="form-group">
+                <label for="selfEvalInstructions">Self-Evaluation instructions for students</label>
+                <textarea ckeditor="editorOptions" rows="8"
```
Can you make this work like `peer_feedback_prompt`

andrew-gardener commented on compair/static/modules/assignment/assignment-module.js:
```diff
@@ -673,21 +739,34 @@ module.factory( "AssignmentPermissions", function (){
-                    'needsSelfEval'     : assignment.status.comparisons.left == 0 && assignment.self_evaluation_needed, // && canCompare (added below)
+                    'isSelfEvalPeriod'  : assignment.self_eval_period,
+                    'canSelfEval'       : assignment.status.answers.answered, // more below
+                    'needsSelfEval'     : assignment.self_eval_period && assignment.self_evaluation_needed && !canManageAssignment && isAfterAnswer,
```
Need to handle when there is no self-eval start/end (so comparisons for the user must be completed first)

andrew-gardener replied:
Your right, sorry I missed it

wynnset commented on compair/api/assignment.py:
```diff
@@ -148,17 +151,38 @@ def post(self, course_uuid, assignment_uuid):
-        # validate answer + comparison period start & end times
+
+        assignment.enable_self_evaluation = params.get('enable_self_evaluation', assignment.enable_self_evaluation)
+        if assignment.enable_self_evaluation:
```
I always program defensively by default, but no need here as you said

wynnset commented on alembic/versions/bbf7d7f7da06_add_assignment_self_eval_dates_and_.py:
```diff
@@ -0,0 +1,32 @@
+    op.add_column('assignment',
+                  sa.Column('self_eval_instructions', sa.Text(), nullable=True))
+
+    op.get_bind().execute("UPDATE assignment SET self_eval_start = compare_end")
```
You're right, I don't know what my logic was for I adding it in. I'll just remove it

wynnset commented:
@andrew-gardener could you expand on what needs to be done in answer-comments api endpoint?

andrew-gardener commented:
@wynnset actually I realize we aren't doing this for answers right now. I will make it a separate issue

lenglund commented:
How are we handling when answers are available to students with this update? My assumption would be the user can see other answers after comparing (or after comparison period ends, if manually set) and prior to self-evaluating regardless if dates set or not.

lenglund commented:
Wondering if we can take a different approach to including the self-eval. Here is what I’m thinking:
* Have radio buttons in the “Assignment Details” area that explicitly have the user choose between with include or not include (with the latter selected by default)
* Then when the user gets to the “Schedule” area, we show another break/checkbox beneath the manually set comparisons dates that is manually setting the self-eval dates (if self-eval selected)
* Then we rename “Comparison Prompts” -> “Prompts” and again add the self-eval instructions below peer feedback instructions with option to set manually (if self-eval selected)

This will also make the comparison count that occurs later in the “Comparison Settings” make more sense (as it includes the self-eval).

lenglund commented:
Terrible screenshot but to give an idea:

![assignment-self-eval](https://user-images.githubusercontent.com/6118592/40191885-8f1673e6-59b7-11e8-9d6b-0766b240ed67.png)

lenglund commented:
Instructional text updates then would be:

Schedule -
> Students can submit and edit their answers only during the answering dates and times you set below. Students do not see an assignment in the application until the answer period begins. Students can read all answers only after the answer period ends and they have completed the required number of comparisons.
> By default, students can compare answers after the answer period ends and self-evaluate (if enabled) after comparing. You may instead set specific dates and times for the comparison and self-evaluation periods. Note that if you manually set a comparison period, students can read all answers only after the answer and comparison periods end.

Prompts - 
> These criteria and instructions appear when students compare answers and self-evaluate (if enabled) for this assignment, to guide them in their critical-thinking process. For each comparison, students have to choose a preferred answer for each criterion listed below and provide peer feedback to both answers as prompted below. For the self-evaluation, students provide reflection to their own answer as prompted below.
> Note that the number of criteria is permanently set once students start comparing, so you won't be able to add or drop criteria after comparing begins. For multiple criteria, if you use the adaptive answer selection setting (see next section), you may also add a weight for each criterion before comparing begins.

wynnset commented:
I've updated the logic, added tests, and also updated the edit screen structure as per @lenglund. Here are the screenshots: 

**Assignment Edit Screen with Self-Eval Enabled:**
![assignment-edit-with-self-eval](https://user-images.githubusercontent.com/8981258/40381908-d52326ec-5db1-11e8-90f9-b90da1e76ea7.png)

**Assignment Edit Screen with Self-Eval _Disabled_:**
![assignment-edit-without-self-eval](https://user-images.githubusercontent.com/8981258/40381909-d53bbbb2-5db1-11e8-99a3-e13910a2bb1e.png)

@andrew-gardener could you review again?

wynnset commented:
Looks like some tests are failing on master, therefore the same tests are failing here too.

andrew-gardener commented:
Tests aren't failing on master? (latest run on travis https://travis-ci.org/ubc/compair/builds/380449405)

wynnset commented:
@lenglund / @andrew-gardener could you please review again? I believe I have addressed all your points.

lenglund commented:
Thanks!

- Changes as discussed re: order of phases.
- On course screen: Text showing due dates should not appear once that activity has been completed (so if no buttons/CTA, no due date shown)
- On course/assignment screens: “Finish Evaluating” -> “Finish Self-Eval” to match with other finish button text
- On comparisons screen: The counting of rounds for comparisons needs to only include comparisons now (not include self-eval)
- On self-evaluation screen: Showing the assignment description doesn’t seem to work? And maybe we should change this to “Show my answer again” anyway (with the dropdown revealing user’s own answer)?
- On self-evaluation screen: Can we make minor updates as indicated in mock-up below (for consistency with comparison screen)?
![self evaluate compair learning application 1](https://user-images.githubusercontent.com/6118592/40997930-17114ac0-68bb-11e8-8da7-0aa208804e0d.png)

wynnset commented:
To clarify, these are the new set of requirements:

- Answer and Comparison start / end date restrictions stay as-is
- Self-eval dates can be set independent of whether comparison dates set or not
- Self-eval can only be done after answering in any given scenario
- Self-eval end date can be set to anytime after the self-eval start

- _If_ the self-eval dates are not explicitly set, self-eval start / end restrictions stay as-is (i.e. self-eval available only during the comparison period after comparisons finished)
- _Otherwise_, self-eval start / end restrictions depend on whether a compare start / end is set (as below)

If comparison dates set:
- Self-eval start date can be set to anytime after the compare start
- Self-eval can only happen after comparison completed while it is the comparison period (this restriction does not apply after the comparison period has ended)

If comparison dates not set:
- Self-eval start date  can be set to anytime after answer end
- Self-eval can only happen after comparisons are completed

wynnset commented:
@andrew-gardener all changes are in place except for the following:
> need to bounce requests to the answer_comment api endpoint on create/edit when outside of self_eval_grace period (and add back-end tests)
Can you explain what you mean by this? I'm not clear what needs to be done here.

wynnset commented:
@lenglund I think I've addressed all your points now. The only thing is I could not replicate the issue you mentioned with the help text showing up when the button was not. Could you please review again?

andrew-gardener commented:
@wynnset you need to do something like this https://github.com/ubc/compair/blob/423d0e131c738eec1aef4cbb0b24175ac5163ab7/compair/api/answer.py#L186 for creating and editing a self-evaluation answer_comment

lenglund commented:
To clarify: The help text (i.e., due date) should only show when that button is showing. Since only one button appears at a time, the help text for that button should appear by itself; not multiple text/dates at once.

lenglund commented:
Other than due date issue, looks good!

wynnset commented:
@andrew-gardener in favor of getting this issue closed, I added a different issue to deal with preventing changing the answer_comment type. Could you please do a final review of this PR and hopefully we can merge it in?

andrew-gardener commented on compair/api/assignment.py:
```diff
@@ -150,9 +153,25 @@ def post(self, course_uuid, assignment_uuid):
 
-        # validate answer + comparison period start & end times
+
+        assignment.enable_self_evaluation = params.get('enable_self_evaluation', assignment.enable_self_evaluation)
```
`assignment.enable_self_evaluation = params.get('enable_self_evaluation', assignment.enable_self_evaluation)` is present here and on line 180 (please remove one)

andrew-gardener commented on compair/api/assignment.py:
```diff
@@ -381,10 +400,20 @@ def post(self, course_uuid):
         if new_assignment.compare_end is not None:
             new_assignment.compare_end = dateutil.parser.parse(params.get('compare_end', None))
 
+        new_assignment.enable_self_evaluation = params.get('enable_self_evaluation', False)
```
`new_assignment.enable_self_evaluation = params.get('enable_self_evaluation', False)` is present here and on line 422 (please remove one)

andrew-gardener commented on compair/models/assignment.py:
```diff
@@ -166,6 +169,28 @@ def after_comparing(self):
+
+    @hybrid_property
+    def self_eval_grace(self):
+        now = dateutil.parser.parse(datetime.datetime.utcnow().replace(tzinfo=pytz.utc).isoformat())
```
`self_eval_grace` should use the same logic as `self_eval_period` but with added 60 seconds (still using compare_grace)

andrew-gardener commented on compair/static/modules/assignment/assignment-module.js:
```diff
@@ -1270,7 +1376,9 @@ module.controller("AssignmentWriteController",
             'cstart': {'date': new Date(), 'time': new Date().setHours(23, 59, 0, 0)},
-            'cend': {'date': new Date(), 'time': new Date().setHours(23, 59, 0, 0)}
+            'cend': {'date': new Date(), 'time': new Date().setHours(23, 59, 0, 0)},
+            'sestart': {'date': new Date(), 'time': new Date().setHours(23, 59, 0, 0)},
```
Can you change the time to `new Date().setHours(0, 0, 0, 0)` for `sestart` and `cstart` (this can close #793)

andrew-gardener commented on compair/static/modules/comparison/comparison-module.js:
```diff
@@ -235,7 +235,12 @@ module.controller(
                                 } else if ($scope.assignment.enable_self_evaluation && ret.status.answers.answered) {
                                     Toaster.success("Comparison Submitted", "Write a self-evaluation now, and your assignment will be complete.");
```
I think this section needs to change a bit
```
} else if ($scope.assignment.enable_self_evaluation && ret.status.answers.answered && !ret.status.comparisons.self_evaluation_completed) {
    $scope.preventExit = false; //user has saved comparison, does not need warning when leaving page
    if ($scope.assignment.self_eval_period) {
        Toaster.success("Comparison Submitted", "Write a self-evaluation now, and your assignment will be complete.");
        $location.path('/course/' + $scope.courseId + '/assignment/' + $scope.assignmentId + '/self_evaluation');
    } else {
        Toaster.success("Comparison Submitted", "...something about waiting for the self-eval period to start");
        $location.path('/course/' + $scope.courseId);
    }
}
```

wynnset commented:
Make sure to address #764 and #793

wynnset commented:
@andrew-gardener if you're happy with this, feel free to merge it in
